### PR TITLE
Update Charts pod from version 3.6.0 to 4.0.3

### PR DESCRIPTION
**Pull request by mokagio**
This update was long due and heralded by https://github.com/wordpress-mobile/WordPress-iOS/pull/18520.

I don't know what the future brings for this library and how we use it, given the new [Swift Charts](https://developer.apple.com/documentation/charts/) framework in the long run (the framework is iOS 16.0+ only).

Regardless, this upgraded is a necessary step to add Xcode 14.0 support (see https://github.com/wordpress-mobile/WordPress-iOS/pull/18833).

image

Even after upgrading to version 4.0.3, I still got the compilation error in Xcode 14.0 beta 1. I got the Charts project itself checked out locally, and it built fine. So, I ended up switch from CocoaPods to SPM and it worked.

We're committed to SPM in the long run, and this change goes in that direction.

## Testing

_WIP: I plan to run the installable build and post some screenshots._

## Regression Notes

1. Potential unintended areas of impact – N.A.
2. What I did to test those areas of impact (or what existing automated tests I relied on) – N.A.
3. What automated tests I added (or what prevented me from doing so) – N.A.

---

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes. **N.A.**
- [x] I have considered adding accessibility improvements for my changes. **N.A.**
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary. **N.A.**

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Stats/Charts/Charts+AxisFormatters.swift [hunk 1]
```diff
@@ -18,7 +18,7 @@ import Charts
 /// The workaround employed here (recommended in 2410 above) relies on the time series data being ordered, and simply
 /// transforms the adjusted values by the time interval associated with the first date in the series.
 ///
-class HorizontalAxisFormatter: IAxisValueFormatter {
+class HorizontalAxisFormatter: AxisValueFormatter {
```
Comment (mokagio) [inline WordPress/Classes/ViewRelated/Stats/Charts/Charts+AxisFormatters.swift:21]: I was disappointed by the Charts releases on GitHub: I couldn't find an upgrade guide.

Luckily, GitHub can show us [the diff between the versions](https://github.com/danielgindi/Charts/compare/v3.6.0...v4.0.3). Add that to the browser's search feature and you can find a guide for these changes.

File: WordPress/Classes/ViewRelated/Stats/Charts/PeriodChart.swift [hunk 2]
```diff
@@ -114,23 +114,21 @@ private final class PeriodChartDataTransformer {
...
 
         let visitorsChartData = BarChartData(dataSet: visitorsDataSet)
         chartData.append(visitorsChartData)
 
-        let likesChartData = BarChartData(entries: likeEntries, valueFormatter: dataSetValueFormatter)
+        let likesChartData = BarChartData(entries: likeEntries)
         chartData.append(likesChartData)
 
-        let commentsChartData = BarChartData(entries: commentEntries, valueFormatter: dataSetValueFormatter)
+        let commentsChartData = BarChartData(entries: commentEntries)
```
Comment (mokagio) [inline WordPress/Classes/ViewRelated/Stats/Charts/PeriodChart.swift]: ⚠️ I've gone back and forth on these and I still need to finish upgrading. The current version would introduce a regression.

File: WordPress/Classes/ViewRelated/Stats/Charts/Charts+Support.swift [hunk 3]
```diff
@@ -4,15 +4,15 @@ import Charts
...
+        let dataSet = BarChartDataSet(entries: entries, valueFormatter: valueFormatter)
         self.init(dataSets: [dataSet])
     }
 }
 
 extension BarChartDataSet {
-    convenience init(values: [BarChartDataEntry], label: String?, valueFormatter: IValueFormatter?) {
-        self.init(entries: values, label: label)
+    convenience init(entries: [BarChartDataEntry], label: String = "", valueFormatter: ValueFormatter) {
+        self.init(entries: entries, label: label)
```
Comment (mokagio) [inline WordPress/Classes/ViewRelated/Stats/Charts/Charts+Support.swift:15]: These were the only non-trivial changes, both due by the fact that neither `label` nor `valueFormatter` can be `nil` now.

I decided to use `""` as a replacement for `nil` for `label`. As for `valueFormatter`, turns out we never set a `nil` one so it was straightforward to simply make it not-optional.

**Comment by sla8c** [comment]
@mokagio I visually tested on device the puffin team's current stats revamp work as well as existing period bar charts and all looks good to me.  cc @Guarani @frosty 

image

**Comment by mokagio** [comment]
Thank you so much @sla8c!

I finally got onto the installable build last night. I didn't know how to access all the graphs, but here are the screenshots of the vists one that I took, just for an extra check on record:

<table>
<tr>
	<td>
IMG_2467
	</td>
<td>
IMG_2466
	</td>
<td>
IMG_2465
	</td>
<td>
IMG_2464
	</td>
</tr>
</table>

File: WordPress/Classes/ViewRelated/Stats/Charts/Charts+Support.swift (hunk 3 above)
Comment (sla8c) [reply WordPress/Classes/ViewRelated/Stats/Charts/Charts+Support.swift:15]: I looked at this closely and I think `""` as the default for label is the correct choice as well.  Thanks for sharing your thoughts here

**Review by sla8c — APPROVED** [review]
LGTM!

**Comment by frosty** [comment]
Thanks for wrangling this, both!

**Comment by guarani** [comment]
+1 thanks for handling this, @mokagio and @sla8c! ❤️

**Pull request merged**